### ISSUE TITLE
Very slow performance compared to default vfs driver

### ISSUE THREAD

**Issue by dg424** [containers/fuse-overlayfs#321]
Hi,

Using the latest fuse-overlayfs driver in a Docker-in-Docker setup, we see that the performance is really bad compared to vfs. Any ideas ? Here are example runs:

To start the rootless docker daemon:

$ docker run --rm -it --privileged --name dind --user 0 --entrypoint sh docker:dind-rootless
/ # apk add fuse-overlayfs
fetch https://dl-cdn.alpinelinux.org/alpine/v3.14/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.14/community/x86_64/APKINDEX.tar.gz
(1/4) Installing fuse-common (3.10.4-r0)
(2/4) Installing fuse3-libs (3.10.4-r0)
(3/4) Installing fuse3 (3.10.4-r0)
(4/4) Installing fuse-overlayfs (1.7.1-r0)
Executing busybox-1.33.1-r3.trigger
OK: 30 MiB in 62 packages
/ # su -c "export DOCKER_TLS_CERTDIR="" && /usr/local/bin/dockerd-entrypoint.sh --tls=false" rootless

The fuse-overlayfs driver is not there by default, so it needs to be added before starting the daemon.

fuse-overlayfs driver test:

$ docker exec -it dind sh
/ # export DOCKER_HOST=127.0.0.1:2375
/ # docker info
Client:
 Context:    default
 Debug Mode: false

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 0
 Server Version: 20.10.8
 Storage Driver: fuse-overlayfs
 Logging Driver: json-file
 Cgroup Driver: none
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: e25210fe30a0a703442421b0f60afac609f950a3
 runc version: v1.0.1-0-g4144b638
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
  rootless
 Kernel Version: 5.10.43.3-microsoft-standard-WSL2
 Operating System: Alpine Linux v3.13 (containerized)
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 15.63GiB
 Name: 9ab4cfd06c57
 ID: N7RM:WEIF:47FB:TF4O:TD3D:CXRJ:23VY:Z4GA:RAM3:XYMJ:QZ5H:ZWOT
 Docker Root Dir: /home/rootless/.local/share/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Product License: Community Engine

WARNING: API is accessible on http://0.0.0.0:2375 without encryption.
         Access to the remote API is equivalent to root access on the host. Refer
         to the 'Docker daemon attack surface' section in the documentation for
         more information: https://docs.docker.com/go/attack-surface/
WARNING: Running in rootless-mode without cgroups. To enable cgroups in rootless-mode, you need to boot the system in cgroup v2 mode.

/ # time docker run python:3.8.10-alpine3.13 /bin/sh -c "apk --no-cache add --upgrade musl-dev gcc python3 python3-dev c
ython libxml2-utils curl libffi-dev openssl-dev make libxml2-dev libxslt-dev && pip install --no-cache-dir pynacl"
Unable to find image 'python:3.8.10-alpine3.13' locally
3.8.10-alpine3.13: Pulling from library/python
540db60ca938: Pull complete
a7ad1a75a999: Pull complete
5472576d6ead: Pull complete
107b9b7cb26b: Pull complete
0c810b10e5b4: Pull complete
Digest: sha256:aa56aba287f05d891b0a184d36cf1f6926922d78b4bfb7c1691790e52905db63
Status: Downloaded newer image for python:3.8.10-alpine3.13
fetch https://dl-cdn.alpinelinux.org/alpine/v3.13/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.13/community/x86_64/APKINDEX.tar.gz
(1/36) Upgrading musl (1.2.2-r0 -> 1.2.2-r1)
(2/36) Upgrading libcrypto1.1 (1.1.1k-r0 -> 1.1.1l-r0)
(3/36) Upgrading libssl1.1 (1.1.1k-r0 -> 1.1.1l-r0)
(4/36) Installing brotli-libs (1.0.9-r3)
(5/36) Installing nghttp2-libs (1.42.0-r1)
(6/36) Installing libcurl (7.78.0-r0)
(7/36) Installing curl (7.78.0-r0)
(8/36) Installing python3 (3.8.10-r0)
(9/36) Installing cython (0.29.21-r1)
(10/36) Installing libgcc (10.2.1_pre1-r3)
(11/36) Installing libstdc++ (10.2.1_pre1-r3)
(12/36) Installing binutils (2.35.2-r1)
(13/36) Installing libgomp (10.2.1_pre1-r3)
(14/36) Installing libatomic (10.2.1_pre1-r3)
(15/36) Installing libgphobos (10.2.1_pre1-r3)
(16/36) Installing gmp (6.2.1-r0)
(17/36) Installing isl22 (0.22-r0)
(18/36) Installing mpfr4 (4.1.0-r0)
(19/36) Installing mpc1 (1.2.0-r0)
(20/36) Installing gcc (10.2.1_pre1-r3)
(21/36) Installing linux-headers (5.7.8-r0)
(22/36) Installing pkgconf (1.7.3-r0)
(23/36) Installing libffi-dev (3.3-r2)
(24/36) Installing zlib-dev (1.2.11-r3)
(25/36) Installing xz-dev (5.2.5-r0)
(26/36) Installing libxml2 (2.9.12-r0)
(27/36) Installing libxml2-dev (2.9.12-r0)
(28/36) Installing libxml2-utils (2.9.12-r0)
(29/36) Installing libgpg-error (1.41-r0)
(30/36) Installing libgcrypt (1.8.8-r0)
(31/36) Installing libxslt (1.1.34-r0)
(32/36) Installing libxslt-dev (1.1.34-r0)
(33/36) Installing make (4.3-r0)
(34/36) Installing musl-dev (1.2.2-r1)
(35/36) Installing openssl-dev (1.1.1l-r0)
(36/36) Installing python3-dev (3.8.10-r0)
Executing busybox-1.32.1-r6.trigger
Executing ca-certificates-20191127-r5.trigger
OK: 246 MiB in 68 packages
Collecting pynacl
  Downloading PyNaCl-1.4.0.tar.gz (3.4 MB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
    Preparing wheel metadata: started
    Preparing wheel metadata: finished with status 'done'
Collecting six
  Downloading six-1.16.0-py2.py3-none-any.whl (11 kB)
Collecting cffi>=1.4.1
  Downloading cffi-1.14.6.tar.gz (475 kB)
Collecting pycparser
  Downloading pycparser-2.20-py2.py3-none-any.whl (112 kB)
Building wheels for collected packages: pynacl, cffi
  Building wheel for pynacl (PEP 517): started
  Building wheel for pynacl (PEP 517): still running...
  Building wheel for pynacl (PEP 517): still running...
  Building wheel for pynacl (PEP 517): still running...
  Building wheel for pynacl (PEP 517): finished with status 'done'
  Created wheel for pynacl: filename=PyNaCl-1.4.0-cp38-cp38-linux_x86_64.whl size=267898 sha256=8b9491dbaa24eaa966ced825d8e0655ebc04b3c7a666107faf79e16be72a6b8f
  Stored in directory: /tmp/pip-ephem-wheel-cache-9lvjkx6f/wheels/10/23/c0/f42041fbad88c47b268b6119919130b407b48e24b41a349684
  Building wheel for cffi (setup.py): started
  Building wheel for cffi (setup.py): finished with status 'done'
  Created wheel for cffi: filename=cffi-1.14.6-cp38-cp38-linux_x86_64.whl size=183884 sha256=9971637015a2cc96b53059a23dcf27c58196c7239d7b178fec7bb2d21ffd9935
  Stored in directory: /tmp/pip-ephem-wheel-cache-9lvjkx6f/wheels/21/a9/81/c074a48686fe8c1ffada1f9e5f53b553c0f766503c0b991c23
Successfully built pynacl cffi
Installing collected packages: pycparser, six, cffi, pynacl
Successfully installed cffi-1.14.6 pycparser-2.20 pynacl-1.4.0 six-1.16.0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
WARNING: You are using pip version 21.1.3; however, version 21.2.4 is available.
You should consider upgrading via the '/usr/local/bin/python -m pip install --upgrade pip' command.
real    4m 52.47s
user    0m 0.07s
sys     0m 0.14s

$ docker run --rm -it --privileged --name dind --user 0 --entrypoint sh docker:dind-rootless
/ # su -c "export DOCKER_TLS_CERTDIR="" && /usr/local/bin/dockerd-entrypoint.sh --tls=false" rootless

vfs - MUCH FASTER:

$ docker exec -it dind sh
/ # export DOCKER_HOST=127.0.0.1:2375
/ # docker info
Client:
 Context:    default
 Debug Mode: false

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 0
 Server Version: 20.10.8
 Storage Driver: vfs
 Logging Driver: json-file
 Cgroup Driver: none
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: e25210fe30a0a703442421b0f60afac609f950a3
 runc version: v1.0.1-0-g4144b638
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
  rootless
 Kernel Version: 5.10.43.3-microsoft-standard-WSL2
 Operating System: Alpine Linux v3.13 (containerized)
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 15.63GiB
 Name: 160f14486e34
 ID: ADOH:MESJ:UTVN:X7WI:HM4Y:KOR4:FBIY:X7MC:6JFI:INQS:GKJH:66AF
 Docker Root Dir: /home/rootless/.local/share/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Product License: Community Engine

WARNING: API is accessible on http://0.0.0.0:2375 without encryption.
         Access to the remote API is equivalent to root access on the host. Refer
         to the 'Docker daemon attack surface' section in the documentation for
         more information: https://docs.docker.com/go/attack-surface/
WARNING: Running in rootless-mode without cgroups. To enable cgroups in rootless-mode, you need to boot the system in cgroup v2 mode.

/ # time docker run python:3.8.10-alpine3.13 /bin/sh -c "apk --no-cache add --upgrade musl-dev gcc python3 python3-dev c
ython libxml2-utils curl libffi-dev openssl-dev make libxml2-dev libxslt-dev && pip install --no-cache-dir pynacl"
Unable to find image 'python:3.8.10-alpine3.13' locally
3.8.10-alpine3.13: Pulling from library/python
540db60ca938: Pull complete
a7ad1a75a999: Pull complete
5472576d6ead: Pull complete
107b9b7cb26b: Pull complete
0c810b10e5b4: Pull complete
Digest: sha256:aa56aba287f05d891b0a184d36cf1f6926922d78b4bfb7c1691790e52905db63
Status: Downloaded newer image for python:3.8.10-alpine3.13
fetch https://dl-cdn.alpinelinux.org/alpine/v3.13/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.13/community/x86_64/APKINDEX.tar.gz
(1/36) Upgrading musl (1.2.2-r0 -> 1.2.2-r1)
(2/36) Upgrading libcrypto1.1 (1.1.1k-r0 -> 1.1.1l-r0)
(3/36) Upgrading libssl1.1 (1.1.1k-r0 -> 1.1.1l-r0)
(4/36) Installing brotli-libs (1.0.9-r3)
(5/36) Installing nghttp2-libs (1.42.0-r1)
(6/36) Installing libcurl (7.78.0-r0)
(7/36) Installing curl (7.78.0-r0)
(8/36) Installing python3 (3.8.10-r0)
(9/36) Installing cython (0.29.21-r1)
(10/36) Installing libgcc (10.2.1_pre1-r3)
(11/36) Installing libstdc++ (10.2.1_pre1-r3)
(12/36) Installing binutils (2.35.2-r1)
(13/36) Installing libgomp (10.2.1_pre1-r3)
(14/36) Installing libatomic (10.2.1_pre1-r3)
(15/36) Installing libgphobos (10.2.1_pre1-r3)
(16/36) Installing gmp (6.2.1-r0)
(17/36) Installing isl22 (0.22-r0)
(18/36) Installing mpfr4 (4.1.0-r0)
(19/36) Installing mpc1 (1.2.0-r0)
(20/36) Installing gcc (10.2.1_pre1-r3)
(21/36) Installing linux-headers (5.7.8-r0)
(22/36) Installing pkgconf (1.7.3-r0)
(23/36) Installing libffi-dev (3.3-r2)
(24/36) Installing zlib-dev (1.2.11-r3)
(25/36) Installing xz-dev (5.2.5-r0)
(26/36) Installing libxml2 (2.9.12-r0)
(27/36) Installing libxml2-dev (2.9.12-r0)
(28/36) Installing libxml2-utils (2.9.12-r0)
(29/36) Installing libgpg-error (1.41-r0)
(30/36) Installing libgcrypt (1.8.8-r0)
(31/36) Installing libxslt (1.1.34-r0)
(32/36) Installing libxslt-dev (1.1.34-r0)
(33/36) Installing make (4.3-r0)
(34/36) Installing musl-dev (1.2.2-r1)
(35/36) Installing openssl-dev (1.1.1l-r0)
(36/36) Installing python3-dev (3.8.10-r0)
Executing busybox-1.32.1-r6.trigger
Executing ca-certificates-20191127-r5.trigger
OK: 246 MiB in 68 packages
Collecting pynacl
  Downloading PyNaCl-1.4.0.tar.gz (3.4 MB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
    Preparing wheel metadata: started
    Preparing wheel metadata: finished with status 'done'
Collecting cffi>=1.4.1
  Downloading cffi-1.14.6.tar.gz (475 kB)
Collecting six
  Downloading six-1.16.0-py2.py3-none-any.whl (11 kB)
Collecting pycparser
  Downloading pycparser-2.20-py2.py3-none-any.whl (112 kB)
Building wheels for collected packages: pynacl, cffi
  Building wheel for pynacl (PEP 517): started
  Building wheel for pynacl (PEP 517): still running...
  Building wheel for pynacl (PEP 517): finished with status 'done'
  Created wheel for pynacl: filename=PyNaCl-1.4.0-cp38-cp38-linux_x86_64.whl size=267898 sha256=c2fb3e7b7558a4e7df2243a2d66ad7c139dc228524a0cca7ac42501d70d213cd
  Stored in directory: /tmp/pip-ephem-wheel-cache-5jq39_jt/wheels/10/23/c0/f42041fbad88c47b268b6119919130b407b48e24b41a349684
  Building wheel for cffi (setup.py): started
  Building wheel for cffi (setup.py): finished with status 'done'
  Created wheel for cffi: filename=cffi-1.14.6-cp38-cp38-linux_x86_64.whl size=183884 sha256=54a0993e9155c2f6b565023d0615be4dac8d82b2d9cc1d3ba758463e663bb294
  Stored in directory: /tmp/pip-ephem-wheel-cache-5jq39_jt/wheels/21/a9/81/c074a48686fe8c1ffada1f9e5f53b553c0f766503c0b991c23
Successfully built pynacl cffi
Installing collected packages: pycparser, six, cffi, pynacl
Successfully installed cffi-1.14.6 pycparser-2.20 pynacl-1.4.0 six-1.16.0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
WARNING: You are using pip version 21.1.3; however, version 21.2.4 is available.
You should consider upgrading via the '/usr/local/bin/python -m pip install --upgrade pip' command.
real    2m 59.60s
user    0m 0.06s
sys     0m 0.08s

**Comment by giuseppe** [comment]
could you try using the `volatile` mount flag for the fuse-overlayfs mount?  Does it make any difference?

**Comment by dg424** [comment]
Hi,
Is this how to set the option ?
--storage-driver=fuse-overlayfs --storage-opt="overlay.mountopt=volatile"

**Comment by giuseppe** [comment]
I am afraid that is not possible with Docker: https://github.com/moby/moby/blob/master/daemon/graphdriver/fuse-overlayfs/fuseoverlayfs.go#L152-L154

Would it be possible for you to re-build fuse-overlayfs?

If that is possible you could hardcode volatile=1 with:

```diff
$ git diff
diff --git a/main.c b/main.c
index 61872a1..7e3d824 100644
--- a/main.c
+++ b/main.c
@@ -5516,7 +5516,7 @@ main (int argc, char *argv[])
                         .lowerdir = NULL,
                         .redirect_dir = NULL,
                         .mountpoint = NULL,
-                        .fsync = 1,
+                        .fsync = 0,
                         .noacl = 0,
                         .squash_to_uid = -1,
                         .squash_to_gid = -1,
```

**Comment by dg424** [comment]
Hi Giuseppe,
Sure, let me try that.
Thanks

**Comment by dg424** [comment]
I'm using Ubuntu WSL2 kernel and I'm getting this error when I try buildah:
$ buildah bud -v $PWD:/build/fuse-overlayfs -t fuse-overlayfs -f ./Containerfile.static.ubuntu .
WARN[0000] Error loading container config when searching for local runtime: mkdir /home/user/rundir/libpod: no such file or directory
ERRO[0000] failed to setup From and Bud flags: failed to get container config: mkdir /home/user/rundir/libpod: no such file or directory

**Comment by dg424** [comment]
Manually did mkdir and looks like it's going through now...

**Comment by dg424** [comment]
Did a few runs with fsync=0 in main.c and they are all around this time:

real    3m 37.95s
user    0m 0.09s
sys     0m 0.04s

vfs is still a bit faster, around:

real    2m 50.97s
user    0m 0.08s
sys     0m 0.05s

Anything else I can try ?

**Comment by giuseppe** [comment]
fuse-overlayfs has better performance when there are multiple layers involved and it avoid the VFS cost of copying each layer multiple times.  In your test case though, where there is just a base layer, I don't think fuse-overlayfs can be as fast as VFS.

**Comment by dg424** [comment]
Hi Giuseppe,

I see. I was hoping there was a way to tune or even hardcode some attributes in the source code to get the same performance as we're looking for the same level of performance as vfs/overlay2 as the key reason for looking at fuse-overlayfs is for linux extended attributes support which vfs does not have. And we can't use overlay2 right now because our kernel does not support overlay for rootless/unprivileged users which is only available in Ubuntu or 5.11+.
